In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_excel("result/GSS.xlsx")

In [3]:
df.head()

,year,id_,wrkstat,hrs1,marital,childs,age,educ,sex,race,income,happy,health,life,trust,satjob,ballot
0,2024,1,Working full time,43,Never married,2,33,4 years of college,MALE,Black,"$25,000 or more",Not too happy,Good,.i: Inapplicable,.i: Inapplicable,Moderately satisfied,Ballot b
1,2024,2,Retired,.i: Inapplicable,Never married,0,64,4 years of college,MALE,White,"$25,000 or more",Pretty happy,Good,.i: Inapplicable,Can't be too careful,.i: Inapplicable,Ballot b
2,2024,3,Retired,.i: Inapplicable,Married,0,69,2 years of college,FEMALE,White,"$25,000 or more",Pretty happy,Good,Routine,.i: Inapplicable,.i: Inapplicable,Ballot c
3,2024,4,Working part time,20,Never married,0,19,12th grade,MALE,Other,"$25,000 or more",Pretty happy,Good,Routine,.i: Inapplicable,Very satisfied,Ballot a
4,2024,5,Retired,.i: Inapplicable,Divorced,3,70,1 year of college,FEMALE,White,"$25,000 or more",Not too happy,Good,Routine,.i: Inapplicable,.i: Inapplicable,Ballot c


In [4]:
df.shape

(3309, 17)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3309 entries, 0 to 3308
Data columns (total 17 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   year     3309 non-null   int64 
 1   id_      3309 non-null   int64 
 2   wrkstat  3309 non-null   object
 3   hrs1     3309 non-null   object
 4   marital  3309 non-null   object
 5   childs   3309 non-null   object
 6   age      3309 non-null   object
 7   educ     3309 non-null   object
 8   sex      3309 non-null   object
 9   race     3309 non-null   object
 10  income   3309 non-null   object
 11  happy    3309 non-null   object
 12  health   3309 non-null   object
 13  life     3309 non-null   object
 14  trust    3309 non-null   object
 15  satjob   3309 non-null   object
 16  ballot   3309 non-null   object
dtypes: int64(2), object(15)
memory usage: 439.6+ KB


##### drop all the rows that have missing values (.n, .d, .s) in all the rows
#### More Notes:
* Most important (drop nas): happy, satjob, income, health, life
* Keep with missing: wrkstat, hrs, educ, marital
* everything else: keep as is

##### Fill keep with missing all other columns (inapplicable, etc.) with unknown
##### Age, Hrs, Childs fill in with the median response
## Goal keep as much data as we can have

In [6]:
df["wrkstat"] = df["wrkstat"].replace(['.n:  No answer', '.d:  Do not Know/Cannot Choose',
                       '.s:  Skipped on Web'], np.nan)


In [7]:
df["hrs1"] = df["hrs1"].replace(".i:  Inapplicable", 0)
df["hrs1"] = df["hrs1"].replace([".n:  No answer", ".d:  Do not Know/Cannot Choose",
                                ".s:  Skipped on Web"], np.nan)

In [8]:
df["marital"] = df["marital"].replace([".s:  Skipped on Web", ".n:  No answer"], np.nan)
df["marital"].value_counts()

marital
Married          1363
Never married    1059
Divorced          516
Widowed           249
Separated         107
Name: count, dtype: int64

In [9]:
df["childs"] = df["childs"].replace([".n:  No answer", ".s:  Skipped on Web"], np.nan)
df["childs"].value_counts()

childs
0            1029
2             851
1             484
3             475
4             243
5              96
6              53
8 or more      31
7              16
Name: count, dtype: int64

In [10]:
df["age"] = df["age"].replace(".n:  No answer", np.nan)
df["age"].value_counts()

age
64    72
37    71
55    71
60    70
42    70
      ..
86    11
81    11
84    10
88     3
87     3
Name: count, Length: 72, dtype: int64

In [11]:
df["educ"] = df["educ"].replace([".d:  Do not Know/Cannot Choose", 
                                 ".n:  No answer"], np.nan)
df["educ"].value_counts()

educ
12th grade                    801
4 years of college            672
2 years of college            443
1 year of college             283
6 years of college            242
5 years of college            179
3 years of college            177
8 or more years of college    138
11th grade                    104
10th grade                     69
7 years of college             54
9th grade                      37
8th grade                      26
6th grade                      23
No formal schooling            12
7th grade                      10
3rd grade                       5
5th grade                       4
4th grade                       3
2nd grade                       3
1st grade                       1
Name: count, dtype: int64

### NOTE: cannot choose, is being dropped but this could be non-binary

In [12]:
df["sex"] = df["sex"].replace(['.s:  Skipped on Web', '.n:  No answer', 
                              '.d:  Do not Know/Cannot Choose'], np.nan)
df["sex"].value_counts()

sex
FEMALE    1823
MALE      1467
Name: count, dtype: int64

In [13]:
df["race"] = df["race"].replace(".i:  Inapplicable", np.nan)
df["race"].value_counts()

race
White    2280
Black     572
Other     392
Name: count, dtype: int64

In [14]:
df["income"] = df["income"].replace([".i:  Inapplicable", ".r:  Refused", ".s:  Skipped on Web"], np.nan)
df["income"].value_counts()

income
$25,000 or more       2266
$10,000 to $14,999     175
$20,000 to $24,999     139
$15,000 to $19,999     103
Under $1,000            67
$1,000 to $2,999        48
$8,000 to $9,999        34
$3,000 to $3,999        26
$5,000 to $5,999        25
$4,000 to $4,999        18
$7,000 to $7,999        16
$6,000 to $6,999        14
Name: count, dtype: int64

In [15]:
df["happy"] = df["happy"].replace([".d:  Do not Know/Cannot Choose", 
                                   ".s:  Skipped on Web", 
                                   ".n:  No answer"], np.nan)
df["happy"].value_counts()

happy
Pretty happy     1892
Not too happy     705
Very happy        684
Name: count, dtype: int64

In [16]:
df["health"] = df["health"].replace([".d:  Do not Know/Cannot Choose", 
                                   ".s:  Skipped on Web", 
                                   ".n:  No answer"], np.nan)
df["health"].value_counts()

health
Good         1755
Fair          808
Excellent     581
Poor          150
Name: count, dtype: int64

In [17]:
df["life"] = df["life"].replace([".d:  Do not Know/Cannot Choose", 
                                   ".s:  Skipped on Web", 
                                   ".n:  No answer",".i:  Inapplicable" ], np.nan)
df["life"].value_counts()

life
Routine     1231
Exciting     845
Dull         144
Name: count, dtype: int64

In [18]:
df["trust"] = df["trust"].replace([".d:  Do not Know/Cannot Choose", 
                                   ".n:  No answer",".i:  Inapplicable" ], np.nan)
df["trust"].value_counts()

trust
Can't be too careful          603
Most people can be trusted    236
Depends                       107
Name: count, dtype: int64

In [19]:
df["satjob"] = df["satjob"].replace([".d:  Do not Know/Cannot Choose", 
                                   ".n:  No answer",".i:  Inapplicable", ".s:  Skipped on Web"], np.nan)
df["satjob"].value_counts()

satjob
Very satisfied           993
Moderately satisfied     922
A little dissatisfied    215
Very dissatisfied        114
Name: count, dtype: int64

In [20]:
df = df.dropna(subset=["happy", "satjob", "income", "health", "life"])

In [21]:
df.columns

Index(['year', 'id_', 'wrkstat', 'hrs1', 'marital', 'childs', 'age', 'educ',
       'sex', 'race', 'income', 'happy', 'health', 'life', 'trust', 'satjob',
       'ballot'],
      dtype='object')

In [22]:
categorical_cols = ["wrkstat", "marital", "trust", "race", "educ", "sex", "income"]
# do not fill in age, hours, childs, but everything else yes.

for col in categorical_cols:
    df[col] = df[col].fillna("Unknown")

In [23]:
## clean age 89 or older (will just be 89)
df["age"] = df["age"].replace("89 or older", 89)

In [24]:
df["age"] = pd.to_numeric(df["age"])

In [25]:
df["hrs1"]= df["hrs1"].replace("89+ hrs", 89)
df["hrs1"] = pd.to_numeric(df["hrs1"])

In [26]:
df["childs"] = df["childs"].replace("8 or more", 8)
df["childs"] = pd.to_numeric(df["childs"])

In [27]:
df["age"] = df["age"].fillna(df["age"].median())
df["hrs1"] = df["hrs1"].fillna(df["hrs1"].median())
df["childs"] = df["childs"].fillna(df["childs"].median())

In [28]:
df.reset_index(drop = True, inplace = True)

In [29]:
df.shape

(1367, 17)

##### Turn the categorical variables into numeric rating system

In [30]:
work_map = {"Working full time": 1, "Working part time": 2, 
            "With a job, but not at work because of temporary illness, vacation, strike": 3, 
           "Unemployed, laid off, looking for work": 4, "Keeping house": 5, "Unknown": 6}

df["wrkstat_n"] = df["wrkstat"].replace(work_map)
df["wrkstat_n"].value_counts() ## no longer use wrkstat use wrkstat_n

wrkstat_n
1    904
2    192
5    154
4     77
3     39
6      1
Name: count, dtype: int64

Note: with highest year of school completed, there's lots of different options. What if we group in together: None, elementary, middle school, high school, some college, college complete, past college (?)

In [31]:
df[["sex", "marital", "educ", "race"]].value_counts()

sex      marital        educ                        race   
FEMALE   Married        4 years of college          White      57
MALE     Married        4 years of college          White      49
                        12th grade                  White      48
         Never married  12th grade                  White      47
FEMALE   Married        12th grade                  White      42
                                                               ..
MALE     Widowed        6 years of college          Unknown     1
                        8 or more years of college  White       1
                        No formal schooling         White       1
Unknown  Married        11th grade                  Other       1
         Never married  3rd grade                   White       1
Name: count, Length: 270, dtype: int64

In [32]:
df[df["sex"] == "Unknown"]
df["trust"].value_counts() ## problem
## there is unknown for trust.
# drop trust column... L
df = df.drop(columns = ["trust"])

In [33]:
df["life"].value_counts()

life
Routine     781
Exciting    505
Dull         81
Name: count, dtype: int64

In [34]:
# what to do with variables that shouldn't be ranked like race, sex, age, etc. 
df.head()

income_map = {"Under $1,000": 1, "$1,000 to $2,999": 2, "$3,000 to $3,999": 3,
             "$4,000 to $4,999": 4, "$5,000 to $5,999": 5, "$6,000 to $6,999": 6, 
             "$7,000 to $7,999": 7, "$8,000 to $9,999": 8, "$10,000 to $14,999": 9,
             "$15,000 to $19,999": 10, "$20,000 to $24,999": 11, "$25,000 or more": 12}

df["income_n"] = df["income"].replace(income_map)

happy_map = {"Very happy": 1, "Pretty happy": 2, "Not too happy": 3}
health_map = {"Excellent": 1, "Good": 2, "Fair": 3, "Poor": 4}
trust_map = {"Most people can be trusted": 1, "Can't be too careful": 2, "Depends": 3}
life_map = {"Exciting": 1, "Routine": 2, "Dull": 3}
work_sat_map = {"Very satisfied": 1, "Moderately satisfied": 2, "A little dissatisfied": 3, "Very dissatisfied": 4}

df["happy_n"] = df["happy"].replace(happy_map)
df["health_n"] = df["health"].replace(health_map)
#df["trust_n"] = df["trust"].replace(trust_map)
df["life_n"] = df["life"].replace(life_map)
df["satjob_n"] = df["satjob"].replace(work_sat_map)
df.head()

,year,id_,wrkstat,hrs1,marital,childs,age,educ,sex,race,...,health,life,satjob,ballot,wrkstat_n,income_n,happy_n,health_n,life_n,satjob_n
0,2024,4,Working part time,20.0,Never married,0.0,19.0,12th grade,MALE,Other,...,Good,Routine,Very satisfied,Ballot a,2,12,2,2,2,1
1,2024,9,Working full time,50.0,Married,2.0,60.0,2 years of college,FEMALE,Black,...,Good,Routine,Moderately satisfied,Ballot c,1,12,1,2,2,2
2,2024,15,Working full time,60.0,Divorced,3.0,63.0,12th grade,FEMALE,White,...,Fair,Exciting,Very satisfied,Ballot a,1,10,1,3,1,1
3,2024,16,Working full time,37.0,Married,1.0,31.0,7 years of college,FEMALE,White,...,Excellent,Exciting,Moderately satisfied,Ballot a,1,12,1,1,1,2
4,2024,17,Working full time,55.0,Married,2.0,64.0,4 years of college,MALE,White,...,Good,Routine,Moderately satisfied,Ballot c,1,12,2,2,2,2


#### How to handle non-ranked data
##### Columns: Marital, Educ, sex, race

In [35]:
df["educ"].value_counts()

educ
12th grade                    320
4 years of college            279
2 years of college            191
1 year of college             122
6 years of college            107
5 years of college             83
8 or more years of college     65
3 years of college             64
11th grade                     47
10th grade                     28
7 years of college             21
9th grade                      11
8th grade                       9
6th grade                       6
No formal schooling             3
Unknown                         3
3rd grade                       2
7th grade                       2
5th grade                       2
2nd grade                       1
4th grade                       1
Name: count, dtype: int64

In [36]:
df["marital"].value_counts()

marital
Married          569
Never married    518
Divorced         185
Separated         49
Widowed           41
Unknown            5
Name: count, dtype: int64

In [37]:
df["race"].value_counts() # race doesn't matter

race
White      905
Black      264
Other      173
Unknown     25
Name: count, dtype: int64

In [38]:
print(df.columns)

Index(['year', 'id_', 'wrkstat', 'hrs1', 'marital', 'childs', 'age', 'educ',
       'sex', 'race', 'income', 'happy', 'health', 'life', 'satjob', 'ballot',
       'wrkstat_n', 'income_n', 'happy_n', 'health_n', 'life_n', 'satjob_n'],
      dtype='object')


In [39]:
print(df.columns)

Index(['year', 'id_', 'wrkstat', 'hrs1', 'marital', 'childs', 'age', 'educ',
       'sex', 'race', 'income', 'happy', 'health', 'life', 'satjob', 'ballot',
       'wrkstat_n', 'income_n', 'happy_n', 'health_n', 'life_n', 'satjob_n'],
      dtype='object')


In [40]:
# EDUC mapping
educ_map = {
    "No formal schooling": 0,
    "2nd grade": 2,
    "3rd grade": 3,
    "4th grade": 4,
    "5th grade": 5,
    "6th grade": 6,
    "7th grade": 7,
    "8th grade": 8,
    "9th grade": 9,
    "10th grade": 10,
    "11th grade": 11,
    "12th grade": 12,
    "1 year of college": 13,
    "2 years of college": 14,
    "3 years of college": 15,
    "4 years of college": 16,
    "5 years of college": 17,
    "6 years of college": 18,
    "7 years of college": 19,
    "8 or more years of college": 20
}

df = df[df['educ'] != 'Unknown']
df['educ'] = df['educ'].map(educ_map)

# MARITAL cleanup
df = df[df['marital'] != 'Unknown']

# One-hot encoding
df = pd.get_dummies(df, columns=['marital'], drop_first=True)
df.head()

,year,id_,wrkstat,hrs1,childs,age,educ,sex,race,income,...,wrkstat_n,income_n,happy_n,health_n,life_n,satjob_n,marital_Married,marital_Never married,marital_Separated,marital_Widowed
0,2024,4,Working part time,20.0,0.0,19.0,12,MALE,Other,"$25,000 or more",...,2,12,2,2,2,1,False,True,False,False
1,2024,9,Working full time,50.0,2.0,60.0,14,FEMALE,Black,"$25,000 or more",...,1,12,1,2,2,2,True,False,False,False
2,2024,15,Working full time,60.0,3.0,63.0,12,FEMALE,White,"$15,000 to $19,999",...,1,10,1,3,1,1,False,False,False,False
3,2024,16,Working full time,37.0,1.0,31.0,19,FEMALE,White,"$25,000 or more",...,1,12,1,1,1,2,True,False,False,False
4,2024,17,Working full time,55.0,2.0,64.0,16,MALE,White,"$25,000 or more",...,1,12,2,2,2,2,True,False,False,False


#### Logistical Regression

Hypothesis: the features that will predict happiness the best are income, sat_job_n, health_n
Features we should look at:
* income_n
* health_n
* satjob_n
* life_n
* wrkstat_n
* age
* childs
* hrs1
* educ
* 

In [45]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import SelectKBest, f_classif, RFE
from sklearn.metrics import accuracy_score, f1_score, classification_report

In [47]:
marital_cols = [col for col in df.columns if col.startswith('marital_')]

feature_cols = [
    'income_n',
    'health_n',
    'satjob_n',
    'life_n',
    'wrkstat_n',
    'age',
    'childs',
    'hrs1',
    'educ'
] + marital_cols

In [49]:
model_df = df[feature_cols + ['happy_n']].dropna()
X = model_df[feature_cols]
y = model_df['happy_n']

print("Features being used:")
print(feature_cols)
print("\nShape of X:", X.shape)
print("Shape of y:", y.shape)

Features being used:
['income_n', 'health_n', 'satjob_n', 'life_n', 'wrkstat_n', 'age', 'childs', 'hrs1', 'educ', 'marital_Married', 'marital_Never married', 'marital_Separated', 'marital_Widowed']

Shape of X: (1359, 13)
Shape of y: (1359,)


In [50]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

log_model = LogisticRegression(max_iter=1000, random_state=42)
log_model.fit(X_train_scaled, y_train)

y_pred_log = log_model.predict(X_test_scaled)

print("\n--- Logistic Regression ---")
print("Accuracy:", accuracy_score(y_test, y_pred_log))
print("F1 Score (weighted):", f1_score(y_test, y_pred_log, average='weighted'))
print(classification_report(y_test, y_pred_log))


--- Logistic Regression ---
Accuracy: 0.6102941176470589
F1 Score (weighted): 0.5554769329695736
              precision    recall  f1-score   support

           1       0.54      0.22      0.32        58
           2       0.62      0.90      0.73       157
           3       0.63      0.21      0.32        57

    accuracy                           0.61       272
   macro avg       0.60      0.44      0.45       272
weighted avg       0.60      0.61      0.56       272



#### Decision Tree:

In [51]:
tree_model = DecisionTreeClassifier(random_state=42, max_depth=5)
tree_model.fit(X_train, y_train)

y_pred_tree = tree_model.predict(X_test)

print("\n--- Decision Tree ---")
print("Accuracy:", accuracy_score(y_test, y_pred_tree))
print("F1 Score (weighted):", f1_score(y_test, y_pred_tree, average='weighted'))
print(classification_report(y_test, y_pred_tree))


--- Decision Tree ---
Accuracy: 0.5992647058823529
F1 Score (weighted): 0.5415794096200333
              precision    recall  f1-score   support

           1       0.48      0.24      0.32        58
           2       0.62      0.89      0.73       157
           3       0.53      0.16      0.24        57

    accuracy                           0.60       272
   macro avg       0.54      0.43      0.43       272
weighted avg       0.57      0.60      0.54       272



### Random Forest

In [52]:
rf_model = RandomForestClassifier(
    n_estimators=100,
    random_state=42,
    max_depth=7
)
rf_model.fit(X_train, y_train)

y_pred_rf = rf_model.predict(X_test)

print("\n--- Random Forest ---")
print("Accuracy:", accuracy_score(y_test, y_pred_rf))
print("F1 Score (weighted):", f1_score(y_test, y_pred_rf, average='weighted'))
print(classification_report(y_test, y_pred_rf))



--- Random Forest ---
Accuracy: 0.6029411764705882
F1 Score (weighted): 0.5238166321285986
              precision    recall  f1-score   support

           1       0.69      0.16      0.25        58
           2       0.60      0.94      0.73       157
           3       0.53      0.14      0.22        57

    accuracy                           0.60       272
   macro avg       0.61      0.41      0.40       272
weighted avg       0.61      0.60      0.52       272



### SelectKBest:

In [53]:
k = min(8, X_train.shape[1])
selector = SelectKBest(score_func=f_classif, k=k)
X_train_kbest = selector.fit_transform(X_train, y_train)
X_test_kbest = selector.transform(X_test)

selected_kbest_features = X.columns[selector.get_support()]
print("\n--- SelectKBest Features ---")
print(list(selected_kbest_features))

# Optional: run logistic regression on SelectKBest features
scaler_k = StandardScaler()
X_train_kbest_scaled = scaler_k.fit_transform(X_train_kbest)
X_test_kbest_scaled = scaler_k.transform(X_test_kbest)

log_kbest = LogisticRegression(max_iter=1000, random_state=42)
log_kbest.fit(X_train_kbest_scaled, y_train)
y_pred_log_kbest = log_kbest.predict(X_test_kbest_scaled)

print("\n--- Logistic Regression with SelectKBest Features ---")
print("Accuracy:", accuracy_score(y_test, y_pred_log_kbest))
print("F1 Score (weighted):", f1_score(y_test, y_pred_log_kbest, average='weighted'))
print(classification_report(y_test, y_pred_log_kbest))


--- SelectKBest Features ---
['health_n', 'satjob_n', 'life_n', 'wrkstat_n', 'hrs1', 'educ', 'marital_Married', 'marital_Never married']

--- Logistic Regression with SelectKBest Features ---
Accuracy: 0.6176470588235294
F1 Score (weighted): 0.5636684978553534
              precision    recall  f1-score   support

           1       0.58      0.24      0.34        58
           2       0.62      0.90      0.74       157
           3       0.63      0.21      0.32        57

    accuracy                           0.62       272
   macro avg       0.61      0.45      0.46       272
weighted avg       0.61      0.62      0.56       272



### RFE

In [54]:
n_features = min(6, X_train.shape[1])

rfe = RFE(
    estimator=LogisticRegression(max_iter=1000, random_state=42),
    n_features_to_select=n_features
)

rfe.fit(X_train_scaled, y_train)
selected_rfe_features = X.columns[rfe.support_]

print("\n--- RFE Selected Features ---")
print(list(selected_rfe_features))

# Run logistic regression on RFE-selected features
X_train_rfe = X_train[selected_rfe_features]
X_test_rfe = X_test[selected_rfe_features]

scaler_rfe = StandardScaler()
X_train_rfe_scaled = scaler_rfe.fit_transform(X_train_rfe)
X_test_rfe_scaled = scaler_rfe.transform(X_test_rfe)

log_rfe = LogisticRegression(max_iter=1000, random_state=42)
log_rfe.fit(X_train_rfe_scaled, y_train)
y_pred_log_rfe = log_rfe.predict(X_test_rfe_scaled)

print("\n--- Logistic Regression with RFE Features ---")
print("Accuracy:", accuracy_score(y_test, y_pred_log_rfe))
print("F1 Score (weighted):", f1_score(y_test, y_pred_log_rfe, average='weighted'))
print(classification_report(y_test, y_pred_log_rfe))


--- RFE Selected Features ---
['health_n', 'satjob_n', 'life_n', 'wrkstat_n', 'hrs1', 'marital_Married']

--- Logistic Regression with RFE Features ---
Accuracy: 0.6213235294117647
F1 Score (weighted): 0.5574193608741206
              precision    recall  f1-score   support

           1       0.59      0.17      0.27        58
           2       0.62      0.93      0.74       157
           3       0.68      0.23      0.34        57

    accuracy                           0.62       272
   macro avg       0.63      0.44      0.45       272
weighted avg       0.63      0.62      0.56       272



In [55]:
feature_importance = pd.Series(rf_model.feature_importances_, index=X.columns)
feature_importance = feature_importance.sort_values(ascending=False)

print("\n--- Random Forest Feature Importance ---")
print(feature_importance)


--- Random Forest Feature Importance ---
health_n                 0.133201
age                      0.128918
life_n                   0.108298
educ                     0.108112
satjob_n                 0.107083
hrs1                     0.098943
childs                   0.082631
marital_Married          0.071884
income_n                 0.058660
wrkstat_n                0.048770
marital_Never married    0.031033
marital_Separated        0.012120
marital_Widowed          0.010348
dtype: float64


## Analysis:

#### Best Model: Logistical Regression (with Feature Selection)
##### Model Comparison:
* Logistic Regression performed best among the baseline models:
    * Accuracy: 61.0%
    * Weighted F1-score: 0.56
* Decision Tree performed slightly lower:
    * Accuracy: 59.9%
    * Weighted F1-score: 0.54
* Random Forest had similar accuracy but lower F1-score:
    * Accuracy: 60.3%
    * Weighted F1-score: 0.52

###### After feature selection:
* Logistic Regression + SelectKBest improved to 61.8% accuracy
* Logistic Regression + RFE was the best overall model:
    * Accuracy: 62.1%
    * Weighted F1-score: 0.56

##### What the Model Performance Suggests:
* Logistical Regression outperforming the tree models suggests that
    * The relationship between the selected features and happiness is mostly linear
    * The data does not contain enough complex non-linear patterns for tree-based models to outperform simpler models
    * Logisitical Regression provided the best balance

##### Common Patterns across Models:
* All models performed the best on the majority happiness class (class 2)
    * Recall for class 2 ("Pretty Happy") was around 0.90-0.94
* All models struggled with predicting the minority classes (class 1 and 3)
    * Lower recall for both extreme happiness groups
* Class Imbalance impacted model preformance
    * Happiness at the extremes may be influenced by more complex or unmeasured factors